Generator/VisitFakeQuery.php

```diff
@@ -0,0 +1,357 @@
+
+        $keywords = ['books','reading','pages','story','fiction','author'];
+        $medium = ['ppc','website','email'];
+        $source = ['bing','google','doubleclick'];
+        for ($i = 0; $i < 50;$i++) {
+
+            $id = bin2hex(random_bytes(3));
+            $campaign = [
+                'id' => $id,
+                'content' => '',
+                'group' => '',
+                'keyword' => $keywords[array_rand($keywords)],
+                'medium' => $medium[array_rand($medium)],
+                'name' => 'campaign_'.$id,
+                'placement' => '',
+                'source' => $source[array_rand($source)]
+            ];
+
+            $this->campaigns[] = $campaign;
+        }
+
+        for ($i = 0; $i < 500;$i++) {
+
+            $id = bin2hex(random_bytes(4));
+            $referer = [
+                'name' => 'ref_'.$id,
+                'url' => 'https://domain'.$id.'.com',
+                'type' => array_rand([1,2,3]),
+                'keyword' => $keywords[array_rand($keywords)],
+            ];
+
+            $this->referers[] = $referer;
+        }
+
+    }
+
+    public function getRandomActionURL(): string
+    {
+
+        // Choose an existing actionsUrl 50% of the time if there are at least 100 in the pool
+        // or 100% of the time if the pool is full
+        if (count($this->actionsUrl) > 100 && (rand(0, 100) < 50 || count($this->actionsUrl) >= $this->actionsPoolSize)) {
+           return $this->actionsUrl[array_rand($this->actionsUrl)];
+        }
+
+        $newActionUrl = "/page_".bin2hex(random_bytes(10));
+        $this->actionsUrl[] = $newActionUrl;
+
+        return $newActionUrl;
+    }
+
+    /**
+     * Get a new visitor id, with a [$chanceReturning] possibility of being a previous visitor
+     *
+     * @param int $chanceReturning
+     *
+     * @return string
+     * @throws \Exception
+     */
+    public function getVisitor(int $chanceReturning): string
+    {
+        if (rand(0,100) <= $chanceReturning && count($this->returnUserIds) > 0) {
+            return $this->returnUserIds[array_rand($this->returnUserIds)];
+        }
+
+        $idvisitor = random_bytes(8);
+
+        // 1% chance to update a return user id if the pool is full
+        if (count($this->returnUserIds) >= $this->returnVisitorPoolSize && rand(0,100) === 1) {
+
+            // Never replace the early return visitors, we want some selects at the quiet end of the index
+            $this->returnUserIds[rand($this->returnVisitorPoolSize / 10, $this->returnVisitorPoolSize)] = $idvisitor;
+        }
+
+        // If ppol is not full then add this id
+        if (count($this->returnUserIds) < $this->returnVisitorPoolSize) {
+            $this->returnUserIds[] = $idvisitor;
+        }
+
+        return $idvisitor;
+    }
+
+    public function getCheckActionExistsQuery(string $actionUrl): array
+    {
+
+        $sql = "SELECT `idaction`, `type`, `name` 
+                FROM log_action 
```

Comment:
@bx80 Why are we not prefixing the `log_action` table here ?

Reply:
Good spot :+1: I've added prefixing for all tables in the class.